src/webui/clientdatastorage.cpp

```diff
@@ -0,0 +1,138 @@
+
+const int CLIENT_DATA_FILE_MAX_SIZE = 1024 * 1024;
+const QString CLIENT_DATA_FILE_NAME = u"web_data.json"_s;
+
+ClientDataStorage::ClientDataStorage(QObject *parent)
+    : QObject(parent)
+{
+    m_clientDataFilePath = specialFolderLocation(SpecialFolder::Data) / Path(CLIENT_DATA_FILE_NAME);
+    if (m_clientDataFilePath.exists())
+    {
+        const auto readResult = Utils::IO::readFile(m_clientDataFilePath, CLIENT_DATA_FILE_MAX_SIZE);
+        if (!readResult)
+        {
+            LogMsg(tr("Failed to load web client data. %1").arg(readResult.error().message), Log::WARNING);
+            return;
+        }
+
+        QJsonParseError jsonError;
+        const QJsonDocument jsonDoc = QJsonDocument::fromJson(readResult.value(), &jsonError);
+        if (jsonError.error != QJsonParseError::NoError)
+        {
+            LogMsg(tr("Failed to parse web client data. File: \"%1\". Error: \"%2\"")
+                .arg(m_clientDataFilePath.toString(), jsonError.errorString()), Log::WARNING);
+            return;
+        }
+
+        if (!jsonDoc.isObject())
+        {
+            LogMsg(tr("Failed to load web client data. File: \"%1\". Error: \"Invalid data format\"")
+                .arg(m_clientDataFilePath.toString()), Log::WARNING);
+            return;
+        }
+
+        m_clientData = jsonDoc.object();
+    }
+}
+
+nonstd::expected<void, QString> ClientDataStorage::storeData(const QJsonObject &object)
+{
+    QJsonObject clientData = m_clientData;
+    bool dataModified = false;
+    for (auto it = object.constBegin(), end = object.constEnd(); it != end; ++it)
+    {
+        const QString &key = it.key();
+        const QJsonValue &value = it.value();
+
+        if (value.isNull())
+        {
+            if (auto it = clientData.find(key); it != clientData.end())
+            {
+                clientData.erase(it);
+                dataModified = true;
+            }
+        }
+        else
+        {
+            const auto &existingValue = clientData.find(key);
+            if (existingValue == clientData.end())
+            {
+                clientData.insert(key, value);
+                dataModified = true;
+            }
+            else if (existingValue.value() != value)
+            {
+                existingValue.value() = value;
+                dataModified = true;
+            }
+        }
+    }
+
+    if (dataModified)
+    {
+        const QByteArray json = QJsonDocument(clientData).toJson(QJsonDocument::Compact);
+        if (json.size() > CLIENT_DATA_FILE_MAX_SIZE)
+            return nonstd::make_unexpected(tr("Total web client data must not be larger than %1 bytes").arg(CLIENT_DATA_FILE_MAX_SIZE));
+        const nonstd::expected<void, QString> result = Utils::IO::saveToFile(m_clientDataFilePath, json);
+        if (!result)
+            return nonstd::make_unexpected(tr("Failed to save web client data. Error: \"%1\"").arg(result.error()));
+
+        m_clientData = clientData;
```

Comment:
I suggest to update `m_clientData` regardless the file saved correctly or not. This way the user can retry saving it later when disk has enough free space and previous changes won't be lost.
```suggestion
        m_clientData = clientData;

        const nonstd::expected<void, QString> result = Utils::IO::saveToFile(m_clientDataFilePath, json);
        if (!result)
            return nonstd::make_unexpected(tr("Failed to save web client data. Error: \"%1\"").arg(result.error()));
```